File: blockchain_go/2_proofofwork/block.go
package main

import (
	"time"
)

type Block struct {
	// 区块头信息 ------
	Timestamp     int64  // 当前时间戳，也就是区块创建的时间
	PrevBlockHash []byte // 前一个 block 的 hash
	Hash          []byte // 当前 block 的 hash
	// 区块头信息 ------
	Data  []byte // 区块存储的实际有效信息，也就是交易
	Nonce int    // 工作量
}

// NewBlock creates and returns Block
func NewBlock(data string, prevBlockHash []byte) *Block {
	block := &Block{
		Timestamp:     time.Now().Unix(),
		PrevBlockHash: prevBlockHash,
		Hash:          []byte{},
		Data:          []byte(data),
		Nonce:         0,
	}
	pow := NewProofOfWork(block)
	nonce, hash := pow.Run()
	// 保存计算结果
	block.Hash = hash[:]
	block.Nonce = nonce // 存储工作量证明

	return block
}

File: blockchain_rust/proofofwork/src/block.rs
use chrono::Utc;

use crate::proofofwork::ProofOfWork;

#[derive(Debug, Clone)]
pub struct Block {
    timestamp: i64,          // 区块创建的时间
    prev_block_hash: String, // 前一个 block 的 hash
    hash: String,            // 当前 block 的 hash
    data: String,            // 区块存储的实际有效信息，也就是交易
    nonce: i64,              // 工作量
}

impl Block {
    pub fn new(data: String, prev_block_hash: String) -> Self {
        let mut block = Self {
            timestamp: Utc::now().timestamp(),
            data,
            prev_block_hash,
            hash: String::new(),
            nonce: 0,
        };

        let pow = ProofOfWork::new(block.clone());
        let (nonce, hash) = pow.run();
        // 保存计算结果
        block.nonce = nonce;
        block.hash = hash;

        block
    }

    /// 创世区块
    pub fn new_genesis_block() -> Self {
        Self::new(String::from("Genesis Block"), String::from(""))
    }

    pub fn get_timestamp(&self) -> i64 {
        self.timestamp
    }

    pub fn get_prev_block_hash(&self) -> String {
        self.prev_block_hash.clone()
    }

    pub fn get_hash(&self) -> String {
        self.hash.clone()
    }

    pub fn get_data(&self) -> String {
        self.data.clone()
    }

    pub fn get_nonce(&self) -> i64 {
        self.nonce
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[ignore]
    #[test]
    fn test_new_block() {
        let block = Block::new(String::from(""), String::from(""));
        println!("{}", block.get_hash());
    }
}
